server/cluster/unsafe_recovery_controller.go

```diff
@@ -823,21 +871,49 @@ func (u *unsafeRecoveryController) buildUpFromReports() (*regionTree, map[uint64
 	return newestRegionTree, peersMap, nil
 }
 
-func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem, forCommitMerge bool) bool {
+func (u *unsafeRecoveryController) selectLeader(peersMap map[uint64][]*regionItem, region *metapb.Region) *regionItem {
+	var leader *regionItem
+	for _, peer := range peersMap[region.GetId()] {
+		if leader == nil || leader.IsRaftStale(peer, u) {
+			leader = peer
+		}
+	}
+	return leader
+}
+
+func (u *unsafeRecoveryController) generateTombstoneTiFlashLearnerPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) bool {
 	if u.err != nil {
 		return false
 	}
 	hasPlan := false
 
-	selectLeader := func(region *metapb.Region) *regionItem {
-		var leader *regionItem
-		for _, peer := range peersMap[region.GetId()] {
-			if leader == nil || leader.IsRaftStale(peer) {
-				leader = peer
+	newestRegionTree.tree.Ascend(func(item btree.Item) bool {
+		region := item.(*regionItem).Region()
+		if !u.canElectLeader(region, false) {
+			leader := u.selectLeader(peersMap, region)
+			if leader == nil {
+				u.err = errors.Errorf("can't select leader for region %v", region)
+				return false
+			}
+			storeID := leader.storeID
+			if u.cluster.GetStore(storeID).IsTiFlash() {
```

Comment:
Only a peer is in TiFlash and selected as a leader will be tombstone?

Reply:
yes, as we can't make Tiflash to be the leader

Reply:
more essentially, Tiflash can't generate snapshot